File: leetcode/java/0125_valid_palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        // loop through s using i (starting at 0) j starting at (s.length() - 1):
        // if (s.charAt(i) == ' ') i++, continue
        // if (s.charAt(j) == ' ') j++, continue
        // if (s.charAt(i).toLower() != s.charAt(j).toLower()) return false
        // i++, j--
        // return true

        s = s.toLowerCase();
        int i = 0, j = s.length() - 1;
        while (i < j) {
            if ((s.charAt(i) < '0' || s.charAt(i) > '9') && (s.charAt(i) < 'a' || s.charAt(i) > 'z')) {
                i++;
                continue;
            }
            if ((s.charAt(j) < '0' || s.charAt(j) > '9') && (s.charAt(j) < 'a' || s.charAt(j) > 'z')) {
                j--;
                continue;
            }
            if (s.charAt(i) != s.charAt(j)) return false;
            i++; 
            j--;
        }

        return true;
    }
}

File: leetcode/cpp/0125_valid_palindrome.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        for (int i = 0, j = s.size() - 1; i <= j; ) {
            if ((tolower(s[i]) < 97 || tolower(s[i]) > 122) && !(tolower(s[i]) >= 48 && tolower(s[i]) <= 57)) {
                i++;
                continue;
            }

            if ((tolower(s[j]) < 97 || tolower(s[j]) > 122) && !(tolower(s[j]) >= 48 && tolower(s[j]) <= 57)) {
                j--;
                continue;
            }

            if (s.size() == 1) {
                return true;
            }

            if (tolower(s[i]) == tolower(s[j])) {
                i++; j--;
            }
            else if (tolower(s[i]) != tolower(s[j])) {
                return false;
            }
        }

        return true;
    }
};